Hatch check phase fails on darwin

### Describe the bug

Hatch doesn't build anymore. Seems a different issue than a year ago https://github.com/NixOS/nixpkgs/issues/209358

### Steps To Reproduce

This started failing again in `1.9.0`.



```
% nix shell nixpkgs#hatch                    
warning: Ignoring setting 'auto-allocate-uids' because experimental feature 'auto-allocate-uids' is not enabled
warning: Ignoring setting 'impure-env' because experimental feature 'configurable-impure-env' is not enabled
error: builder for '/nix/store/4mgjxh6ri2my6x0wfr2pmwfyv7qhnwkh-hatch-1.9.0.drv' failed with exit code 1;
       last 10 log lines:
       > FAILED tests/cli/run/test_run.py::test_sync_dependencies - AssertionError: assert 'Checking dependencies\n' == 'Checking dep...ependen...
       > FAILED tests/python/test_core.py::TestGetInstalled::test_order - hatch.errors.PythonDistributionResolutionError: Could not find a default so...
       > FAILED tests/cli/env/test_create.py::test_features - AssertionError: assert 1 == 2
       > FAILED tests/cli/run/test_run.py::test_sync_project_dependencies - AssertionError: assert 'Checking dependencies\n' == 'Checking dep...ependen...
       > FAILED tests/cli/run/test_run.py::test_sync_project_features - AssertionError: assert 'Checking dependencies\n' == 'Checking dep...ependen...
       > FAILED tests/cli/run/test_run.py::test_dependency_hash_checking - AssertionError: assert 'Checking dependencies\n' == 'Checking dep...ependen...
       > FAILED tests/cli/env/test_create.py::test_sync_dynamic_dependencies - AssertionError: assert 3 == 4
       > FAILED tests/cli/env/test_create.py::test_sync_dependencies - AssertionError: assert 'Creating env...ependencies\n' == 'Creating env...ep...
       > ================= 17 failed, 1628 passed, 55 skipped in 51.96s =================
       > /nix/store/ymqnr6q4slam0ns76mjpcjry70fa8h3i-stdenv-darwin/setup: line 1591: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/4mgjxh6ri2my6x0wfr2pmwfyv7qhnwkh-hatch-1.9.0.drv'.
       ```
```

### Expected behavior
It builds

### Screenshots
If applicable, add screenshots to help explain your problem.

### Additional context

Related to https://github.com/NixOS/nixpkgs/pull/271586 ?

### Notify maintainers

@onny 


---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/ha/hatch/package.nix
+++ b/pkgs/by-name/ha/hatch/package.nix
@@ -8,13 +8,18 @@
 
 python3.pkgs.buildPythonApplication rec {
   pname = "hatch";
-  version = "1.9.0";
+  version = "1.9.7";
   format = "pyproject";
 
   src = fetchPypi {
     inherit pname version;
-    hash = "sha256-4ealEeFS7HzU26vE9Pahh0hwvUnJfRfTkLkjLdpoXOM=";
+    hash = "sha256-Gae4IXn5Tyrd2612qn5lq5DK1VqxA9U4J2N5NcnmYkw=";
   };
+
+  postPatch = ''
+    # Loosen hatchling runtime version dependency
+    sed -i 's/hatchling<1.22/hatchling/' pyproject.toml
+  '';
 
   nativeBuildInputs = with python3.pkgs; [
     hatchling
@@ -69,19 +74,13 @@ python3.pkgs.buildPythonApplication rec {
     "test_editable_exact_force_include_option"
     "test_editable_exact_force_include_build_data_precedence"
     "test_editable_pth"
-    # AssertionError: assert len(extract_installed_requirements(output.splitlines())) > 0
-    "test_creation_allow_system_packages"
-    # cli table output mismatch
-    "test_context_formatting"
     # expects sh, finds bash
     "test_all"
     "test_already_installed_update_flag"
     "test_already_installed_update_prompt"
-    # unmet expectations about the binary module we provide
-    "test_dependency_not_found"
-    "test_marker_unmet"
-    # output capturing mismatch, likely stdout/stderr mixup
-    "test_no_compatibility_check_if_exists"
+    # Loosen hatchling runtime version dependency
+    "test_core"
+    "test_correct"
   ] ++ lib.optionals stdenv.isDarwin [
     # https://github.com/NixOS/nixpkgs/issues/209358
     "test_scripts_no_environment"
@@ -90,6 +89,8 @@ python3.pkgs.buildPythonApplication rec {
     # It is not possible to run it in a nix build using a /nix/store shell.
     # See https://github.com/pypa/hatch/pull/709 for the relevant code.
     "test_populate_default_popen_kwargs_executable"
+  ] ++ lib.optionals stdenv.isAarch64 [
+    "test_resolve"
   ];
 
   meta = with lib; {

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -8932,8 +8932,6 @@ with pkgs;
 
   haste-server = callPackage ../servers/haste-server { };
 
-  hatch = python3Packages.callPackage ../development/tools/hatch { };
-
   hal-hardware-analyzer = libsForQt5.callPackage ../applications/science/electronics/hal-hardware-analyzer {
     stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
   };

